ipv8/attestation/wallet/pengbaorange/structs.py

```diff
@@ -0,0 +1,266 @@
+from __future__ import absolute_import, division
+
+from binascii import hexlify, unhexlify
+from struct import pack, unpack
+
+from six.moves import xrange
+
+from ..pengbaorange.boudot import EL, SQR
+from ..primitives.boneh import decode, encode
+from ..primitives.structs import BonehPublicKey, ipack, iunpack
+from ..primitives.value import FP2Value
+from ...identity_formats import Attestation
+
+
+def _serialize_fp2value(value):
+    normalized = value.wp_compress()
+    return ipack(normalized.a) + ipack(normalized.b)
+
+
+def _unserialize_fp2value(mod, value):
+    rem = value
+    a, rem = iunpack(rem)
+    b, rem = iunpack(rem)
+    return FP2Value(mod, a, b), rem
+
+
+class PengBaoCommitment(object):
+
+    def __init__(self, c, c1, c2, ca, ca1, ca2, ca3, caa):  # pylint: disable=R0913
+        self.c = c
+        self.c1 = c1
+        self.c2 = c2
+        self.ca = ca
+        self.ca1 = ca1
+        self.ca2 = ca2
+        self.ca3 = ca3
+        self.caa = caa
+
+    def serialize(self):
+        return (ipack(self.c.mod) + _serialize_fp2value(self.c) + _serialize_fp2value(self.c1)
+                + _serialize_fp2value(self.c2) + _serialize_fp2value(self.ca) + _serialize_fp2value(self.ca1)
+                + _serialize_fp2value(self.ca2) + _serialize_fp2value(self.ca3) + _serialize_fp2value(self.caa))
+
+    @classmethod
+    def unserialize(cls, s):
+        """
+        :param s: the string to unserialize
+        :type s: str
+        :return: the unserialized commitment
+        :rtype: PengBaoCommitment
+        """
+        mod, rem = iunpack(s)
+        c, rem = _unserialize_fp2value(mod, rem)
+        c1, rem = _unserialize_fp2value(mod, rem)
+        c2, rem = _unserialize_fp2value(mod, rem)
+        ca, rem = _unserialize_fp2value(mod, rem)
+        ca1, rem = _unserialize_fp2value(mod, rem)
+        ca2, rem = _unserialize_fp2value(mod, rem)
+        ca3, rem = _unserialize_fp2value(mod, rem)
+        caa, rem = _unserialize_fp2value(mod, rem)
+        return cls(c, c1, c2, ca, ca1, ca2, ca3, caa), rem
+
+
+class PengBaoCommitmentPrivate(object):
+
+    MSGSPACE = list(range(256))
+
+    def __init__(self, m1, m2, m3, r1, r2, r3):
+        self.m1 = m1
+        self.m2 = m2
+        self.m3 = m3
+        self.r1 = r1
+        self.r2 = r2
+        self.r3 = r3
+
+    def generate_response(self, s, t):
```

Comment:
`s` and `t` cannot be zero, isn't it? A check or assertion somewhere would be good.

Reply:
Not only should `s` and `t` be non-zero, they are also expected to be "large integers". The paper doesn't really specify how large, other than strictly non-zero. I opted for `LARGE_INTEGER`.

I think this also answers your other two comments above. This function is only strictly called after the previous assertion that `s` and `t` are larger than `LARGE_INTEGER`. Also, from the other side, when generating `s` and `t` they are made larger than `LARGE_INTEGER`.

Reply:
I can also add some additional comments to the code about this of course 👍